### PR TITLE
Add basic object-level permissions support

### PR DESCRIPTION
**Pull request by tgoyne**
This adds schema definitions for the permissions object types, exposes the function to get the computed privileges for a Realm, Class or Object, and adds a basic test of the functionality. Still needs documentation.

## ☑️ ToDos

* [x] 📝 Changelog entry
* [x] 🚦 Tests
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [x] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: src/js_realm.hpp [hunk 1]
```diff
@@ -186,6 +185,7 @@ class RealmClass : public ClassDefinition<T, SharedRealm, ObservableClass<T>> {
     static void compact(ContextType, ObjectType, Arguments, ReturnValue &);
     static void delete_model(ContextType, ObjectType, Arguments, ReturnValue &);
     static void object_for_object_id(ContextType, ObjectType, Arguments, ReturnValue&);
+    static void privileges(ContextType, ObjectType, Arguments, ReturnValue&);
```
Comment (kneth) [inline src/js_realm.hpp]: Why no `#if REALM_ENABLE_SYNC`?

File: src/js_realm.hpp [hunk 2]
```diff
@@ -1041,21 +1007,67 @@ void RealmClass<T>::object_for_object_id(ContextType ctx, ObjectType this_object
     if (!sync::has_object_ids(realm->read_group()))
         throw std::logic_error("Realm._objectForObjectId() can only be used with synced Realms.");
 
-    std::string object_type = Value::validated_to_string(ctx, args[0]);
-    validated_object_schema_for_value(ctx, realm, args[0], object_type);
-
+    auto& object_schema = validated_object_schema_for_value(ctx, realm, args[0]);
     std::string object_id_string = Value::validated_to_string(ctx, args[1]);
-    auto object_id = object_id_from_string(object_id_string);
+    auto object_id = sync::ObjectID::from_string(object_id_string);
```
Comment (kneth) [inline src/js_realm.hpp]: Can you access `sync::ObjectID` when `REALM_ENABLE_SYNC` isn't defined?

File: tests/js/permission-tests.js [hunk 3]
```diff
@@ -118,5 +146,70 @@ module.exports = {
...
+
+    testObjectPermissions() {
+      let config = (user, url) => {
+        return {
+          schema: [
+            Realm.Permissions.Permission,
+            Realm.Permissions.User,
+            Realm.Permissions.Role,
+            {
+              name: 'Object',
```
Comment (kneth) [inline tests/js/permission-tests.js]: Maybe a little confusing name since JS also has `Object`.

File: tests/js/permission-tests.js [hunk 4]
```diff
@@ -57,6 +57,34 @@ function repeatUntil(fn, predicate) {
     return check;
 }
 
+function subscribe(results) {
+  const subscription = results.subscribe()
```
Comment (bmunkholm) [inline tests/js/permission-tests.js]: indent if off

File: src/js_realm.hpp (hunk 1 above)
Comment (tgoyne) [reply src/js_realm.hpp]: The object store code takes care of abstracting away sync/non-sync logic (non-sync realms just always report that you have full permissions).

File: src/js_realm.hpp (hunk 2 above)
Comment (tgoyne) [reply src/js_realm.hpp]: This code is guarded by `REALM_ENABLE_SYNC`.

Comment (kneth) [reply src/js_realm.hpp]: Oh yeah, line 1005

**Pull request merged**